py/asmrv32: Fix RV32 extended test failures.

agatti opened the pull request:
### Summary

This PR contains fixes for the native emitter test failures uncovered in #15551.  These fixes depend on #15573 to be merged in to properly fix test some test failures.

### Testing

Ran the whole `esp32` test suite via `mpy-cross` on an ESP32C3 board with `./run-tests.py --target esp32 --device /dev/ttyUSB0 --via-mpy --emit native --mpy-cross-flags="-march=rv32imc"`.

agatti committed:
py/asmrv32: Fix short/long jumps scheduling.

The RV32 emitter always scheduled short jumps even outside the emit
compiler pass.  Running the full test suite through the native emitter
instead of just the tests that depend on the emitter at runtime (as in,
`micropython/native_*` and `micropython/viper_* tests`) uncovered more
places where the invalid behaviour was still present.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented on py/asmrv32.c:
```diff
@@ -326,10 +326,11 @@ void asm_rv32_emit_call_ind(asm_rv32_t *state, mp_uint_t index) {
-    ptrdiff_t displacement = (ptrdiff_t)(state->base.label_offsets[label] - state->base.code_offset);
+    mp_uint_t label_offset = state->base.label_offsets[label];
+    ptrdiff_t displacement = (ptrdiff_t)(label_offset - state->base.code_offset);
+    bool emit_short_jump = (label_offset != (mp_uint_t)-1) && (displacement < 0);
```
I suggest making a local (static inline?) function that does these three lines of computation, and reused, to reduce code duplication.

agatti replied:
Good point!  I'll update my local copy and push this change along with fixes for the other failing tests, thanks.

agatti committed:
py/asmrv32: Emit C.LW opcodes only when necessary.

The RV32 emitter sometimes generated short load opcodes even when it
was not supposed to.  This commit fixes an off-by-one error in its
offset eligibility range calculation and corrects one case of offset
calculation, operating on the raw label index number rather than its
effective offset in the stack (C.LW assumes all loads are
word-aligned).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented on py/asmrv32.c:
```diff
@@ -270,6 +270,15 @@ static void emit_function_epilogue(asm_rv32_t *state, mp_uint_t registers) {
     state->saved_registers_mask = old_saved_registers_mask;
 }
 
+static void calculate_displacement_for_label(asm_rv32_t *state, mp_uint_t label, ptrdiff_t *displacement, bool *emit_short_jump) {
```
I suggest making this return `bool` for the `emit_short_jump` variable, instead of taking that as a pointer-to-return.  That makes the code a little neater, and maybe more efficient because it's not taking the address of a local variable.

Also, I suggest calling the variable `can_emit_short_jump` because it's not a requirement to emit a short jump, just a hint that it's possible.

agatti replied:
The function not returning a `bool` was deliberate, as it's more difficult to forget about the second returned value that way, since it is forced to be declared for the code to compile (and the compiler would reorder everything anyway when optimising things - there are *a lot* of unused registers on RV32 after all).  I agree with the variable renaming, it is definitely clearer that way!

That said, I've implemented your suggestions, thanks for taking a look.

dpgeorge commented on py/asmrv32.c:
```diff
@@ -354,31 +364,25 @@ void asm_rv32_emit_jump_if_reg_eq(asm_rv32_t *state, mp_uint_t rs1, mp_uint_t rs
+    calculate_displacement_for_label(state, label, &displacement, &emit_short_jump);
 
-    if (FIT_SIGNED(displacement, 9) && IS_IN_C_REGISTER_WINDOW(rs)) {
+    if (FIT_SIGNED(displacement, 8) && emit_short_jump && IS_IN_C_REGISTER_WINDOW(rs)) {
```
I suggest to put the `emit_short_jump` (or `can_emit_short_jump`) first in the if-logic, because it's the most important thing to check first, whether this optimisation can be done or not.  It'll also be a little more efficient when executing.

agatti replied:
I've implemented your suggestion, but the compiler emitted the very same code when optimisations were enabled as `can_emit_short_jump` is the least computationally-intensive thing to check for if you need to invalidate the whole expression (a non-volatile boolean), so it was checked first anyway.

dpgeorge replied:
Oh, that's interesting that the compiler swaps the operands of the short-circuiting operator.  I guess it can do that if it's all local variables and arithmetic (no function calls).

Anyway, thanks for updating it.

dpgeorge commented on py/asmrv32.c:
```diff
@@ -445,12 +458,14 @@ void asm_rv32_emit_mov_reg_local(asm_rv32_t *state, mp_uint_t rd, mp_uint_t loca
+    // lui   temporary, HI(offset) ; Or c.lui if possible
+    // c.add temporary, sp
+    // lw    rd, LO(offset)(temporary)
+    load_upper_immediate(state, INTERNAL_TEMPORARY, upper);
```
Why did this change to use `INTERNAL_TEMPORARY` instead of `rd`?  It seems like using `rd` is OK and then it doesn't require an additional register.

agatti replied:
Too eager when copying and pasting I guess.  `INTERNAL_TEMPORARY` is always saved and restored as part of the function prologue/epilogue and its clobbering is always function-scoped so it's safe.  I'll fix that to use `rd` though, thanks.

agatti replied:
There are other places where the temporary register is used, like in `asm_rv32_emit_jump` and `asm_rv32_emit_call_ind` for example.

If I have to pick another register to replace that with, I'd say I could use REG_TEMP2 instead.  I understand that right now is only used inside `emit_native_rot_three` and in viper load/stores for RV32, but I wonder if there are any chances of that register being used elsewhere in the future.  If that's OK for me to use then this may allow few more size optimisations, as `S0` is inside the compressed opcodes register window.

dpgeorge replied:
If using `INTERNAL_TEMPORARY` instead of `rd` allows for more size optimisations, then that's a good reason to use `INTERNAL_TEMPORARY`.

agatti replied:
I've looked a bit more into this to see if by shuffling things around there are more avenues for size optimisation.

Something that can be done (not in this PR, of course), would be to free up the register taken by `INTERNAL_TEMPORARY` (`S0`) and map it to `REG_TEMP0`.  Whilst this will still require a push/pop of the register in question (S-registers have to be saved and restored in this case), it would allow usage of compressed opcodes like `C.SUB`, `C.AND`, `C.OR` and `C.XOR`, and make it viable to use `C.SW` in more places.  There would be a penalty when calling out to the MicroPython runtime as REG_TEMP0 would probably be already clobbered when calling `asm_rv32_emit_call_ind`.

I'd need to come up with some numbers to see if such a thing is justified, though.

Anyway, I'll change the code to use `rd` and `REG_TEMP2` whenever it won't impact on code size.  Hopefully there's a way to make this bit of code better than it is now.

agatti commented:
This PR now contains fixes for all RV32-specific test failures found in #15551, it's now marked as ready for review.

dpgeorge commented:
I tested this PR on a RP2350 in RISC-V mode and it works well, all `basic`, `extmod` and `float` tests now pass (they didn't before):
```
$ ./run-tests.py --target rp2 --via-mpy --emit native --mpy-cross-flags="-march=rv32imc" -d basics extmod float
...
706 tests performed (22699 individual testcases)
706 tests passed
66 tests skipped
```

agatti committed:
py/asmrv32: Use REG_TEMP2 whenever possible.

The RV32 emitter used an additional temporary register, as certain code
sequences required extra storage.  This commit removes its usage in all
but one case, using REG_TEMP2 instead.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
I've cleaned the PR up a bit, and did the same successful test verification you did but on an ESP32C3.

dpgeorge commented:
This looks really good now, thank you.  I also reconfirmed the latest changes on RP2350-RISCV.
